plugins/pixart-rf/fu-pxi-firmware.c

```diff
@@ -25,39 +28,57 @@ fu_pxi_rf_firmware_parse (FuFirmware *firmware,
 			  GError **error)
 {
 	gsize bufsz = 0;
-	guint8 pos = 0;
 	const guint8 *buf;
+	guint8 fw_header[PIXART_RF_FW_HEADER_SIZE];
 	g_autoptr(FuFirmwareImage) img = fu_firmware_image_new (fw);
+	const guint8 TAG[PIXART_RF_FW_HEADER_TAG_SIZE] = {0x55, 0xAA, 0x55, 0xAA, 0x55, 0xAA, 0x55, 0xAA};
+	guint8 fw_header_tag[PIXART_RF_FW_HEADER_TAG_SIZE];
+	gboolean check_header_exist = TRUE;
 
 	/* get buf */
 	buf = g_bytes_get_data (fw, &bufsz);
-	if (bufsz < 32) {
+	if (bufsz < PIXART_RF_FW_HEADER_SIZE) {
 		g_set_error (error,
 			     G_IO_ERROR,
 			     G_IO_ERROR_FAILED,
 			     "firmware invalid, too small!");
 		return FALSE;
 	}
+	/* get fw header */
+	if (!fu_memcpy_safe (fw_header, PIXART_RF_FW_HEADER_SIZE, 0x0,
+			     buf, bufsz, bufsz - PIXART_RF_FW_HEADER_SIZE,
+			     PIXART_RF_FW_HEADER_SIZE, error)) {
+		g_prefix_error (error, "failed to read fw header ");
+		return FALSE;
+	}
 	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL) {
-		fu_common_dump_raw (G_LOG_DOMAIN, "fw last 32 bytes",
-				    &buf[bufsz - 32], 32);
+		fu_common_dump_raw (G_LOG_DOMAIN, "fw header",
+				    fw_header, PIXART_RF_FW_HEADER_SIZE);
+	}
+	
+	/* get fw header tag */
+	if (!fu_memcpy_safe (fw_header_tag, PIXART_RF_FW_HEADER_TAG_SIZE, 0x0,
+			     fw_header, PIXART_RF_FW_HEADER_SIZE, PIXART_RF_FW_HEADER_TAG_OFFSET,
+			     PIXART_RF_FW_HEADER_TAG_SIZE, error)) {
+		g_prefix_error (error, "failed to read fw header tag ");
+		return FALSE;
 	}
 
-	/* find the version tag */
-	for (guint32 idx = 0; idx < 32; idx++) {
-		if (buf[(bufsz - 32) + idx ] == 'v') {
-			pos = idx;
+	/* check the TAG from fw header is correct */
+	for (guint32 idx = 0x0; idx < PIXART_RF_FW_HEADER_TAG_SIZE; idx++) {
```

Comment:
and here; plus doesn't it need to be `sizeof(fw_header) - sizeof(tag)`?

Reply:
No, I have to copy tag data from header array to tag_arrray